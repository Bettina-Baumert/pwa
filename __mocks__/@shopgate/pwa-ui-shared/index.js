export const ProgressBar = {
  hide: jest.fn(),
  show: jest.fn(),
};
