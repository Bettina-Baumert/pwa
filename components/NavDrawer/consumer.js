import { consume } from 'redux-props';
import { historyPush } from '@shopgate/pwa-common/actions/router';

/**
 * @returns {Object}
 */
const mapProps = ({ dispatch }) => ({
  navigate: (pathname, title) => () => dispatch(historyPush({
    pathname,
    state: { title },
  })),
});

/**
 * @returns {boolean}
 */
const stateChanged = () => false;

export default consume({
  mapProps,
  stateChanged,
});
