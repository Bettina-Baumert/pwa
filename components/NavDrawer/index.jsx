/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import I18n from '@shopgate/pwa-common/components/I18n';
import Portal from '@shopgate/pwa-common/components/Portal';
import * as portals from '@shopgate/pwa-common/constants/Portals';
import { INDEX_PATH, PAGE_PATH } from '@shopgate/pwa-common/constants/RoutePaths';
import { CATEGORY_PATH } from '@shopgate/pwa-common-commerce/category/constants';
import { CART_PATH } from '@shopgate/pwa-common-commerce/cart/constants';
import { FAVORITES_PATH } from '@shopgate/pwa-common-commerce/favorites/constants';
import showReturnPolicy from '@shopgate/pwa-common-commerce/market/helpers/showReturnPolicy';
import appConfig from '@shopgate/pwa-common/helpers/config';
import ClientInformation from 'Components/ClientInformation';
import HomeIcon from 'Components/icons/HomeIcon';
import HeartIcon from 'Components/icons/HeartIcon';
import ViewListIcon from 'Components/icons/ViewListIcon';
import ShoppingCartIcon from 'Components/icons/ShoppingCartIcon';
import LocalShippingIcon from 'Components/icons/LocalShippingIcon';
import InfoIcon from 'Components/icons/InfoIcon';
import CreditCardIcon from 'Components/icons/CreditCardIcon';
import DescriptionIcon from 'Components/icons/DescriptionIcon';
import SecurityIcon from 'Components/icons/SecurityIcon';
import LogoutIcon from 'Components/icons/LogoutIcon';
import Layout from './components/Layout';
import Item from './components/Item';
import CartItem from './components/CartItem';
import Divider from './components/Divider';
import Header from './components/Header';
import connect from './connector';

/**
 * The NavDrawer component.
 */
class NavDrawer extends Component {
  static propTypes = {
    toggleNavDrawer: PropTypes.func.isRequired,
    cartProductCount: PropTypes.number,
    entries: PropTypes.shape(),
    highlightFavorites: PropTypes.bool,
    logout: PropTypes.func,
    navDrawerActive: PropTypes.bool,
    user: PropTypes.shape(),
  };

  static defaultProps = {
    cartProductCount: 0,
    entries: {},
    highlightFavorites: false,
    logout: () => {},
    navDrawerActive: false,
    user: null,
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.contentRef = null;
  }

  /**
   * Reset scroll position when drawer opens.
   */
  componentWillReceiveProps({ navDrawerActive }) {
    if (this.contentRef && navDrawerActive && !this.props.navDrawerActive) {
      this.contentRef.scrollTop = 0;
    }
  }

  /**
   * Sets a ref to the content element in order to reset scroll position.
   * @param {HTMLElement} ref The element ref.
   */
  setContentRef = (ref) => {
    this.contentRef = ref;
  };

  /**
   * Handles the close event for the drawer and propagates the changes to the store.
   */
  handleClose = () => {
    this.props.toggleNavDrawer(false);
  };

  /**
   * Renders Items for the given menu entries.
   * @param {Array} entries The menu entries.
   * @returns {JSX}
   */
  renderEntries(entries) {
    return entries.map(entry => (
      <Item key={entry.url} href={entry.url} close={this.handleClose}>
        <I18n.Text string={entry.label} />
      </Item>
    ));
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      entries,
      user,
      cartProductCount,
      navDrawerActive,
      logout,
    } = this.props;

    const showQuickLinks = entries.quicklinks && !!entries.quicklinks.length;

    const props = {
      handleClose: this.handleClose,
      Item,
    };

    return (
      <Layout
        active={navDrawerActive}
        close={this.handleClose}
        setContentRef={this.setContentRef}
      >

        {/* Header */}
        <Portal name={portals.NAV_MENU_HEADER_BEFORE} />
        <Portal
          name={portals.NAV_MENU_HEADER}
          props={{
            ...props,
            user,
          }}
        >
          <Header user={user} close={this.handleClose} />
        </Portal>
        <Portal name={portals.NAV_MENU_HEADER_AFTER} />

        <Portal name={portals.NAV_MENU_CONTENT_BEFORE} />

        {/* Home */}
        <Portal name={portals.NAV_MENU_HOME_BEFORE} />
        <Portal name={portals.NAV_MENU_HOME} props={props}>
          <Item href={INDEX_PATH} icon={HomeIcon} close={this.handleClose}>
            <I18n.Text string="navigation.home" />
          </Item>
        </Portal>
        <Portal name={portals.NAV_MENU_HOME_BEFORE} />

        {/* Categories */}
        <Portal name={portals.NAV_MENU_CATEGORIES_BEFORE} />
        <Portal name={portals.NAV_MENU_CATEGORIES} props={props}>
          <Item href={CATEGORY_PATH} icon={ViewListIcon} close={this.handleClose}>
            <I18n.Text string="navigation.categories" />
          </Item>
        </Portal>
        <Portal name={portals.NAV_MENU_CATEGORIES_BEFORE} />

        {appConfig.hasFavorites && (
          <Fragment>
            <Portal name={portals.NAV_MENU_CATEGORIES_BEFORE} />
            <Portal
              name={portals.NAV_MENU_CATEGORIES}
              props={{
                ...props,
                withIndicator: this.props.highlightFavorites,
              }}
            >
              <Item
                href={FAVORITES_PATH}
                icon={HeartIcon}
                close={this.handleClose}
                withIndicator={this.props.highlightFavorites}
              >
                <I18n.Text string="navigation.favorites" />
              </Item>
            </Portal>
            <Portal name={portals.NAV_MENU_CATEGORIES_BEFORE} />
          </Fragment>
        )}

        <CartItem
          href={CART_PATH}
          icon={ShoppingCartIcon}
          count={cartProductCount}
          close={this.handleClose}
        >
          <I18n.Text string="navigation.cart" />
        </CartItem>

        <Divider close={this.handleClose} />

        {showQuickLinks && this.renderEntries(entries.quicklinks)}
        {showQuickLinks && <Divider close={this.handleClose} />}

        <Item href={`${PAGE_PATH}/shipping`} icon={LocalShippingIcon} close={this.handleClose}>
          <I18n.Text string="navigation.shipping" />
        </Item>

        <Item href={`${PAGE_PATH}/payment`} icon={CreditCardIcon} close={this.handleClose}>
          <I18n.Text string="navigation.payment" />
        </Item>

        <Divider close={this.handleClose} />

        <Item href={`${PAGE_PATH}/terms`} icon={DescriptionIcon} close={this.handleClose}>
          <I18n.Text string="navigation.terms" />
        </Item>

        <Item href={`${PAGE_PATH}/privacy`} icon={SecurityIcon} close={this.handleClose}>
          <I18n.Text string="navigation.privacy" />
        </Item>

        {showReturnPolicy && (
          <Item href={`${PAGE_PATH}/return_policy`} icon={DescriptionIcon} close={this.handleClose}>
            <I18n.Text string="navigation.return_policy" />
          </Item>
        )}

        <Item href={`${PAGE_PATH}/imprint`} icon={InfoIcon} close={this.handleClose}>
          <I18n.Text string="navigation.about" />
        </Item>

        {user && <Divider close={this.handleClose} />}
        {user && (
          <Item onClick={logout} icon={LogoutIcon} close={this.handleClose}>
            <I18n.Text string="navigation.logout" />
          </Item>
        )}
        <Portal name={portals.NAV_MENU_CONTENT_AFTER} />
        <ClientInformation />

      </Layout>
    );
  }
}

export default connect(NavDrawer);

export { NavDrawer as Unwrapped };
