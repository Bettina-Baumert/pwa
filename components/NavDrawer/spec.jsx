import React from 'react';
import { shallow, mount } from 'enzyme';
import Backdrop from '@shopgate/pwa-common/components/Backdrop';
import ConnectedNavDrawer, { Unwrapped as NavDrawer } from './index';
import Item from './components/Item';
import headerStyles from './components/Header/style';

// Mock the portal
jest.mock('react-portal', () => (
  ({ isOpened, children }) => (
    isOpened ? children : null
  )
));

// Mock the parsed link component.
jest.mock('@shopgate/pwa-common/helpers/parsed-link', () => (
  class {
    open = () => {};
  }
));

// Mock the redux connect() method instead of providing a fake store.
jest.mock('./connector', () => obj => obj);
jest.mock('Components/ClientInformation/connector', () => (obj) => {
  const newObj = obj;

  newObj.defaultProps = {
    client: {},
    enableDebugLogging: () => {},
  };

  return newObj;
});

describe('<NavDrawer />', () => {
  let toggleNavDrawerMock;

  beforeEach(() => {
    jest.useFakeTimers();
    toggleNavDrawerMock = jest.fn();
    window.requestAnimationFrame = jest.fn();
  });

  it('should render without optional props', () => {
    const wrapper = shallow(
      <ConnectedNavDrawer toggleNavDrawer={toggleNavDrawerMock} />
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('Item').length).toBeGreaterThan(0);
  });

  it('should render the navigation drawer opened', () => {
    const wrapper = mount(
      <ConnectedNavDrawer
        navDrawerActive
        toggleNavDrawer={toggleNavDrawerMock}
      />
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(Backdrop).length).toBe(1);
  });

  it('should trigger the toggle navigation callback when clicked on backdrop', () => {
    const wrapper = mount(
      <ConnectedNavDrawer
        navDrawerActive
        toggleNavDrawer={toggleNavDrawerMock}
      />
    );

    expect(wrapper).toMatchSnapshot();

    const backdrop = wrapper.find(Backdrop);

    expect(toggleNavDrawerMock).not.toHaveBeenCalled();

    backdrop.find('div[aria-hidden]').simulate('click');

    expect(toggleNavDrawerMock).toHaveBeenCalled();
  });

  it('should trigger the toggle navigation callback when clicked on an item', () => {
    const wrapper = mount(
      <ConnectedNavDrawer
        navDrawerActive
        toggleNavDrawer={toggleNavDrawerMock}
      />
    );

    expect(wrapper).toMatchSnapshot();

    const item = wrapper.find(Item).at(0).find('div[aria-hidden]').at(0);

    expect(toggleNavDrawerMock).not.toHaveBeenCalled();

    // Simulate the click.
    item.simulate('click');
    // Simulate the click delay.
    jest.runAllTimers();

    expect(toggleNavDrawerMock).toHaveBeenCalled();
  });

  it('should render the header properly when a user is logged in', () => {
    const user = {
      firstName: 'Foo',
      mail: 'foo@bar.baz',
    };
    const wrapper = mount(
      <NavDrawer
        user={user}
        toggleNavDrawer={toggleNavDrawerMock}
      />
    );

    expect(wrapper).toMatchSnapshot();

    const headerLoggedInSelector = `.${headerStyles.loggedIn.split(' ').join('.')}`;

    expect(wrapper.find(headerLoggedInSelector).length).toBe(1);
  });

  it('should render the header properly when a user is logged out', () => {
    const wrapper = mount(
      <NavDrawer
        user={null}
        toggleNavDrawer={toggleNavDrawerMock}
      />
    );

    expect(wrapper).toMatchSnapshot();

    const headerLoggedInSelector = `.${headerStyles.loggedIn.split(' ').join('.')}`;

    expect(wrapper.find(headerLoggedInSelector).length).toBe(0);
  });
});
