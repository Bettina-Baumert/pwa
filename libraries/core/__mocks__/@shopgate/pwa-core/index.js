export const UIEvents = {
  emit: jest.fn(),
  on: jest.fn(),
};
