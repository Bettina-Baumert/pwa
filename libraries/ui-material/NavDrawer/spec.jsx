import React from 'react';
import { shallow } from 'enzyme';
import { UIEvents } from '@shopgate/pwa-core';
import NavDrawer from './index';

describe('NavDrawer', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<NavDrawer />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should open and close with an event', () => {
    const wrapper = shallow(<NavDrawer />);

    UIEvents.emit(NavDrawer.OPEN);
    expect(wrapper.state().open).toEqual(true);

    UIEvents.emit(NavDrawer.CLOSE);
    expect(wrapper.state().open).toEqual(false);
  });

  it('should close when Backdrop is clicked', () => {
    const wrapper = shallow(<NavDrawer />);
    const backdrop = wrapper.find('Backdrop');

    backdrop.simulate('click');

    expect(wrapper.state().open).toEqual(false);
  });
});
