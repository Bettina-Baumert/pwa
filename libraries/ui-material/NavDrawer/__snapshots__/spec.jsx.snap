// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavDrawer should match the snapshot 1`] = `
<React.Fragment>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={false}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={300}
    unmountOnExit={false}
  />
  <Backdrop
    className=""
    color="#000"
    duration={200}
    isVisible={false}
    level={4}
    onClick={[Function]}
    opacity={20}
  />
</React.Fragment>
`;
